src/profilers.cpp

```diff
@@ -2,19 +2,373 @@
+
+          std::exit(errno);
+        }
+        else
+        {
+          close(pipe_fd[1]);
+          //read
+          char buffer[1024];
+          ssize_t bytes_read;
+          std::string data;
+          
+          while ((bytes_read = read(pipe_fd[0], buffer, sizeof(buffer) - 1)) > 0) {
+              buffer[bytes_read] = '\0';
+              data += buffer;
+          }
+          std::string parsed_data;
+
+          if(!this->regex.empty()){
+
+            std::regex regexPattern(this->regex);
+            std::smatch match_pattern;
+            if (std::regex_search(data, match_pattern, regexPattern)) {
+                parsed_data = match_pattern.str();
+            }else{
+                std::exit(ERROR_USER_REGEX_MATCH);
+            }
+
+          }else{
+            parsed_data = data;
+          }
+
+          std::regex number_regex(R"(-?\d+(\.\d+)?)");;
+
+          std::smatch match;
+          std::string temp_data = parsed_data;
+          int count = 0;
+          float metric_val;
+                    
+          while (std::regex_search(temp_data, match, number_regex)) {
+            count++;
+            if(count > 1){break;}
+  
+            std::string number_str = match.str(0);
+              try {
+                  metric_val = std::stof(number_str);  
+              } catch (const std::exception&) {
+                    // failed conversion need to handle later
+                    std::exit(ERROR_CONVERSION_TO_FLOAT);
+              }
+            temp_data = match.suffix().str();
+          
+          }
+          if(count == 0){
+            //no numbers found 
+            std::exit(ERROR_NO_NUMBER_REGEX);
+          }
+          if(count > 1){
+            //error too many numbers
+            std::exit(ERROR_TOO_MANY_NUMBERS_REGEX);
+          }
+
+          close(pipe_fd[0]);
+
+
+          // ["<CUSTOM_METRIC>", <metric-reading command string>, <user-provided metric name string>, <timestamp in nanoseconds>, <value of the metric>]
+          clock_gettime(CLOCK_MONOTONIC, &end);
+          long timestamp = (end.tv_sec - start.tv_sec) * 1000 * 1000000 + (end.tv_nsec - start.tv_nsec);
+          nlohmann::json metric = nlohmann::json::array({"<CUSTOM_METRIC>", this->metric_command , this->metric_name, timestamp, metric_val });
+
+          std::string s = metric.dump();
+
+
+          //connection write s
+          std::unique_ptr<Connection> connection;
+
+          std::regex pipe_regex(R"(pipe\s+(\d+)_(\d+))");
+          std::regex tcp_regex(R"(TCP\s+([\d\.]+)_(\d+))");
+          std::smatch instruction_match;
+
+          if (std::regex_match(instrs, instruction_match, pipe_regex)) {
+              
+              int read_fd[2];
+              int write_fd[2];
+
+              write_fd[1] = std::stoi(match[1].str());
+              read_fd[0] = std::stoi(match[2].str());
+              connection = std::make_unique<FileDescriptor>(write_fd, read_fd, this->server_buffer);
```

Comment:
I've forgotten to tell you that you should also send a special message to indicate to the other end that you're trying to connect to it, see how the Python script does it: https://github.com/AdaptivePerf/AdaptivePerf/blob/non-sampling-based/src/scripts/adaptiveperf-process.py#L80-L83.

Reply:
Also, the order of file descriptors is incorrect.

It should be:
```cpp
read_fd[0] = std::stoi(match[1].str());
write_fd[1] = std::stoi(match[2].str());
connection = std::make_unique<FileDescriptor>(read_fd, write_fd, this->server_buffer);
```

(I'll probably simplify ```FileDescriptor``` so that only two file descriptors rather than 4 should be provided as 2 of these are not actually used)

Reply:
I can see that in the python script the "connect" message is only sent if it's a pipe (not a socket). Is that correct?

Reply:
Yes, this is only in case of a pipe (because it doesn't have a notion of connection by itself, unlike a TCP socket).